### PR TITLE
remove remote cli dependency for CLI

### PR DESCRIPTION
**Pull request by arturrez**
sdk PR to follow
ref: https://github.com/ava-labs/avalanche-tooling-sdk-go/issues/76

### CODE REVIEW

File: cmd/nodecmd/wiz.go [hunk 1]
```diff
@@ -73,6 +73,7 @@ var (
 	deployTeleporterMessenger              bool
 	deployTeleporterRegistry               bool
 	replaceKeyPair                         bool
+	depricatedRemoteCLIVersion             string
```
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/nodecmd/wiz.go:76]: it would force us to update hypersdk example for morpheusVM

[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/nodecmd/wiz.go:76]: k. will do

File: cmd/nodecmd/sync.go [hunk 2]
```diff
@@ -78,34 +84,63 @@ func syncSubnet(_ *cobra.Command, args []string) error {
 	return nil
 }
 
+// prepareSubnetPlugin creates subnet plugin to all nodes in the cluster
+func prepareSubnetPlugin(hosts []*models.Host, subnetName string) error {
+	sc, err := app.LoadSidecar(subnetName)
+	if err != nil {
+		return err
+	}
+	ux.Logger.Info("Creating subnet %s plugin for hosts: %s", subnetName, hosts)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/nodecmd/sync.go:93]: fixed

File: pkg/ssh/installer.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+type HostInstaller struct {
+	Host *models.Host
+}
+
+func NewHostInstaller(host *models.Host) *HostInstaller {
+	return &HostInstaller{Host: host}
+}
+
+func (h *HostInstaller) GetArch() (string, string) {
+	goArhBytes, err := h.Host.Command("go env GOARCH", nil, constants.SSHScriptTimeout)
```
Comment (felipemadero) [inline pkg/ssh/installer.go:21]: I wonder if we can avoid depending on go binary for this

File: pkg/ssh/ssh.go [hunk 4]
```diff
@@ -502,42 +494,253 @@ func RunSSHUploadStakingFiles(host *models.Host, nodeInstanceDirPath string) err
...
+	// make sure that genesis and bootstrap data is preserved
+	if genesisFileExists(host) {
+		avagoConf.GenesisPath = filepath.Join(constants.DockerNodeConfigPath, constants.GenesisFileName)
+	}
+	remoteAvagoConfFile, err := getAvalancheGoConfigData(host)
+	if err != nil {
 		return err
 	}
-	if _, err := host.Command(fmt.Sprintf("/home/ubuntu/bin/avalanche subnet join %s %s --avalanchego-config /home/ubuntu/.avalanchego/configs/node.json --plugin-dir /home/ubuntu/.avalanchego/plugins --force-write", subnetName, networkFlag), nil, constants.SSHScriptTimeout); err != nil {
+	bootstrapIDs, _ := utils.GetValueString(remoteAvagoConfFile, "bootstrap-ids")
```
Comment (felipemadero) [inline pkg/ssh/ssh.go:527]: we should check for error here

File: pkg/ssh/ssh.go [hunk 5]
```diff
@@ -502,42 +494,253 @@ func RunSSHUploadStakingFiles(host *models.Host, nodeInstanceDirPath string) err
...
+	bootstrapIPs, _ := utils.GetValueString(remoteAvagoConfFile, "bootstrap-ips")
```
Comment (felipemadero) [inline pkg/ssh/ssh.go:528]: same

File: pkg/ssh/ssh.go [hunk 6]
```diff
@@ -502,42 +494,253 @@ func RunSSHUploadStakingFiles(host *models.Host, nodeInstanceDirPath string) err
...
+		installURL, _, err := dl.GetDownloadURL(sc.VMVersion, hostInstaller) // extension is tar.gz
+		if err != nil {
+			return err
+		}
+
+		archiveName := "subnet-evm.tar.gz"
+		archiveFullPath := filepath.Join(tmpDir, archiveName)
+
+		// download and install subnet evm
+		if _, err := host.Command(fmt.Sprintf("%s %s -O %s", "busybox wget", installURL, archiveFullPath), nil, constants.SSHLongRunningScriptTimeout); err != nil {
```
Comment (felipemadero) [inline pkg/ssh/ssh.go:589]: I believe you have scripts for this on shell, eg upgradeSubnetEVM, pkg/ssh/shell/getNewSubnetEVMRelease.sh, why not using them?

Comment (felipemadero) [reply pkg/ssh/ssh.go:589]: why not having one .sh file for all of this?

File: pkg/ssh/ssh.go [hunk 7]
```diff
@@ -502,42 +494,253 @@ func RunSSHUploadStakingFiles(host *models.Host, nodeInstanceDirPath string) err
...
+		}
+		defer os.Remove(subnetConfigFile.Name())
+		if err := os.WriteFile(subnetConfigFile.Name(), subnetConfig, constants.WriteReadUserOnlyPerms); err != nil {
+			return err
+		}
+		subnetConfigPath := filepath.Join(constants.CloudNodeConfigPath, "subnets", subnetIDStr+".json")
+		if err := host.MkdirAll(filepath.Dir(subnetConfigPath), constants.SSHDirOpsTimeout); err != nil {
+			return err
+		}
+		if err := host.Upload(subnetConfigFile.Name(), subnetConfigPath, constants.SSHFileOpsTimeout); err != nil {
```
Comment (felipemadero) [inline pkg/ssh/ssh.go:689]: can we have an upload that takes as input a slice of bytes, so we dont need to create tmp files?

File: pkg/ssh/ssh.go [hunk 8]
```diff
@@ -702,3 +905,31 @@ func composeFileExists(host *models.Host) bool {
...
+
+func getAvalancheGoConfigData(host *models.Host) (map[string]interface{}, error) {
+	// get remote node.json file
+	nodeJSONPath := filepath.Join(constants.CloudNodeConfigPath, constants.NodeFileName)
+	tmpFile, err := os.CreateTemp("", "avalanchecli-node-*.json")
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(tmpFile.Name())
+	if err := host.Download(nodeJSONPath, tmpFile.Name(), constants.SSHFileOpsTimeout); err != nil {
```
Comment (felipemadero) [inline pkg/ssh/ssh.go:922]: can we have a download that returns a byte slice?

File: pkg/utils/common.go [hunk 9]
```diff
@@ -559,3 +559,11 @@ func Command(cmdLine string, params ...string) *exec.Cmd {
 	c.Env = os.Environ()
 	return c
 }
+
+// GetValueString returns the value of a key in a map as a string.
+func GetValueString(data map[string]interface{}, key string) (string, error) {
```
Comment (felipemadero) [inline pkg/utils/common.go:564]: can we improve the name of the function?

File: pkg/utils/json.go [hunk 10]
```diff
@@ -25,3 +25,14 @@ func ValidateJSON(path string) ([]byte, error) {
 
 	return contentBytes, nil
 }
+
+// MergeJsonMaps merges two maps of type map[string]interface{}
+func MergeJSONMaps(a, b map[string]interface{}) map[string]interface{} {
```
Comment (felipemadero) [inline pkg/utils/json.go:30]: can be maps.Copy be used instead of MergeJSONMaps?

File: pkg/ssh/installer.go (hunk 3 above)
Comment (arturrez) [reply pkg/ssh/installer.go:21]: good point. agree

File: pkg/ssh/ssh.go (hunk 7 above)
Comment (arturrez) [reply pkg/ssh/ssh.go:689]: agree. we should have another func for this

File: pkg/ssh/ssh.go (hunk 8 above)
Comment (arturrez) [reply pkg/ssh/ssh.go:922]: agree. we should have another func for this

File: pkg/utils/common.go (hunk 9 above)
Comment (arturrez) [reply pkg/utils/common.go:564]: which one do you suggest ?

File: pkg/ssh/ssh.go (hunk 6 above)
Comment (arturrez) [reply pkg/ssh/ssh.go:589]: it will increase size of `scriptInputs` struct so I though we can get away without it as operation is straightforward

**Pull request merged**